binrw_derive/src/codegen/typed_builder.rs

```diff
@@ -0,0 +1,248 @@
+use proc_macro2::TokenStream;
+use quote::{quote, ToTokens};
+use syn::{Ident, Type};
+
+#[allow(clippy::wildcard_imports)]
+use crate::codegen::sanitization::*;
+use crate::parser::meta_types::NamedImport;
+
+pub(crate) enum BuilderFieldKind {
+    Required,
+    Optional { default: Box<syn::Expr> },
+}
+
+pub(crate) struct BuilderField {
+    pub(crate) name: Ident,
+    pub(crate) ty: Type,
+    pub(crate) kind: BuilderFieldKind,
+}
+
+pub(crate) struct Builder<'a> {
+    pub(crate) builder_name: &'a Ident,
+    pub(crate) result_name: &'a Ident,
+    pub(crate) fields: &'a [BuilderField],
+}
+
+impl<'a> Builder<'a> {
+    pub(crate) fn generate(&self) -> TokenStream {
+        let builder_name = self.builder_name;
+        let name = self.result_name;
+        let fields = self.generate_result_fields();
+        let builder_fields = self.generate_builder_fields();
+        let initial = self.generate_builder_initial();
+        let generics = self.generate_generics();
+        let initial_generics = self.generate_initial_generics();
+        let setters = self.generate_setters();
+        let satisfied = std::iter::repeat(SATISFIED_OR_OPTIONAL);
+        let field_names: Vec<_> = self.fields.iter().map(|field| &field.name).collect();
+        let possible_unwrap = self.fields.iter().map(BuilderField::possible_unwrap);
+        quote!(
+            #[derive(Clone)]
+            pub(crate) struct #name {
+                #fields
+            }
+
+            impl #name {
+                pub fn builder() -> #builder_name < #( #initial_generics ),* > {
```

Comment:
Uhhhhh nope. I wrote this all by hand, didn't even really reference any other code. Do you mean the spaces? those were mostly because it was a bit unreadable all packed together